### ISSUE TITLE
Option to turn specific errors into warnings (e.g. don't fail if a declared variable isn't used)

### ISSUE THREAD

**Issue by d4h0** [TypeStrong/ts-node#1896]
### Desired Behavior

Hi,

It would be great if we could turn specific diagnostics from being fatal (i.e. failing compilation/execution) into warnings (they are printed to the terminal, but the script gets executed).

### Is this request related to a problem?

Yes. It's pretty painful if a script can't be executed because, for example, a declared variable isn't used (TS6133). Fixing non-fatal diagnostics like that often leads to significant refactoring work.

I think, it would be better, to just warn about these issues. This makes it easier to experiment with stuff.

### Alternatives you've considered

There doesn't seem to be an alternative that I like.

Some potential options are:

- Turn off type checking completely, or for specific issues (e.g. `ts-node --transpileOnly` or `ts-node --ignoreDiagnostics <code,code>`)
- Don't fail compilation, just print a warning (e.g. `ts-node --logError`).

### Additional context

Basically, what I'm asking for is `ts-node --logError <code,code>`

**Comment by cspotcode** [comment]
Not sure if the added complexity and maintenance burden is warranted given that many tools are already available.

`--logError` already makes diagnostics non-fatal yet still visible.

TS already has `@ts-ignore` and `@ts-expect-error`.

Unused variables can already be prefixed with an underscore to suppress TS6133.

**Comment by d4h0** [comment]
> Not sure if the added complexity and maintenance burden is warranted given that many tools are already available.

I'm not familiar with the code base, but it seems unlikely that this would increase complexity or maintenance burden significantly.

As mentioned, this could be implemented by extending `--logError` to accept specific diagnostic codes (i.e. `--logError <code,code>`). `--ignoreDiagnostics` already has the option to filter by diagnostic codes, so adding that functionality to `--logError` would increase consistency, IMO.

Everything needed, seems to be implemented already, therefore I would expect that only a few lines of code would need to be added or changed (but as mentioned, I'm not familiar with the code).

IMO, defaulting to failing if a variable isn't used is a huge mistake (the same for just ignoring it). Workarounds don't change much. DX is significantly decreased.

The Rust compiler is an example, for a tool that defaults to warning users if a variable isn't used (I would get insane if Rust would default to failing).

Personally, I've just turned this diagnostic off in `ts-node`, and hope that I will catch unused variables in my editor (where the diagnostic isn't turned off). But I don't think that is an ideal solution.

**Comment by cspotcode** [comment]
It sounds like you're more bothered by typescript's diagnostic behavior than by ts-node, since we use typescript and surface the diagnostics it gives us.  Typescript has options to enable and disable those checks.  You can also search their issue tracker for similar requests; this has probably been discussed.

**Comment by d4h0** [comment]
> Typescript has options to enable and disable those checks.

That is exactly what I've done (as mentioned above). The problem is, that disabled checks disable the diagnostic completely (i.e. no warning is displayed).

> It sounds like you're more bothered by typescript's diagnostic behavior than by ts-node

I don't think so. As far as I understand, it is `ts-node` that is throwing the exception if TypeScript returns an 'Variable not used' diagnostic.

If that is incorrect, then you are indeed right, and I should report this somewhere else.

Anyway, I strongly feel that failing on non-fatal issues like unused variables is bad for UX, and I also feel I was pretty clear why I think this way.

So if you don't think what I've mentioned above is an issue (or if the issue is indeed somewhere else), feel free to close this issue. If the above doesn't convince you, then I doubt that anything else I could say, would.

**Comment by cspotcode** [comment]
I'm thinking of TS's own `noUnused*` options.  There's probably some relevant discussion on TS issue tracker about them.
